src/dbnode/persist/fs/msgpack/decoder.go

```diff
@@ -425,15 +467,48 @@ func (dec *Decoder) decodeIndexEntry(bytesPool pool.BytesPool) schema.IndexEntry
 	actualChecksum := dec.readerWithDigest.digest().Sum32()
 
 	// Decode checksum field originally added in V3
-	expectedChecksum := uint32(dec.decodeVarint())
-
-	if expectedChecksum != actualChecksum {
+	v := dec.decodeVarint()
+	indexEntry.IndexChecksum = v
+	if indexEntry.IndexChecksum != int64(actualChecksum) {
 		dec.err = errorIndexEntryChecksumMismatch
 	}
 
 	return indexEntry
 }
 
+func (dec *Decoder) decodeIndexChecksum(
+	compareID []byte,
+	bytesPool pool.BytesPool,
+) (int64, IndexChecksumLookupStatus) {
+	entry := dec.decodeIndexEntry(bytesPool)
+	if dec.err != nil {
+		return 0, NotFound
+	}
+
+	compare := bytes.Compare(compareID, entry.ID)
+	var checksum int64
+	if compare == 0 {
+		// NB: need to compute hash before freeing entry bytes.
+		checksum = dec.hasher.HashIndexEntry(entry)
+	}
+
+	// free elements if pooled.
+	if bytesPool != nil {
+		bytesPool.Put(entry.ID)
+		bytesPool.Put(entry.EncodedTags)
+	}
+
+	if compare > 0 {
+		// seekID should appear after the current entry.
```

Comment:
Can you reword this comment to "compareID" appears after "entry.ID" to make it explicit? "seekID" and "current entry" is ambiguous a little.